File: 3213-count-subarrays-where-max-element-appears-at-least-k-times/solution.java
class Solution {
    public long countSubarrays(int[] nums, int k) {
        int n = nums.length;
        int i=0;
        int j=0;
        int maxElt = 0;
        for(int elt=0; elt<n; elt++) {
            maxElt = Math.max(maxElt, nums[elt]);//3
        }
        int maxCount = 0;
        long count = 0;
        
        while(j < n) {
            if(nums[j] == maxElt) {
                maxCount++;
            }

            while(maxCount >= k) {
                if(nums[i] == maxElt) {
                    maxCount--;
                }
                i++;
            }
            count += i;
            j++;
        }
        return count;
    }
}

/*

💡 What does “ending at j” mean?

It means:

We fix the right boundary j,
and count how many left boundaries are valid.

⸻

🧠 Why do we shrink?

We shrink to find:

The smallest left index where condition still holds.

Once we go one step further, condition breaks.

That boundary tells us:

All starts before this point are valid.

*/

File: 3213-count-subarrays-where-max-element-appears-at-least-k-times/solution.swift
class Solution {
    func countSubarrays(_ nums: [Int], _ k: Int) -> Int {
        var i=0
        var j=0
        var n = nums.count
        var maxElt = nums.max()!
        // DICT is NOT required
        var dict = [Int:Int]()
        var count = 0
        while j < n {
            dict[nums[j], default:0] += 1
            
            while dict[maxElt, default:0] >= k {
                // atleast >= k, always 0 to i subarrays are valid
                count += n-j
                dict[nums[i]]! -= 1
                if dict[nums[i]] == 0 {
                    dict.removeValue(forKey: nums[i])
                } 
                i += 1
            }
            j += 1
        }
        return count
    }
}
